lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -1395,6 +1395,33 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
     return nullptr;
   }
 
+  template <typename DeclType> bool isFilePrivate(DeclType *D) {
+    return (D->hasAccess() && D->getFormalAccess() == AccessLevel::FilePrivate);
+  }
+
+  /// The private discriminator is represented as an inline namespace.
+  llvm::DIScope *getFilePrivateScope(llvm::DIScope *Parent, TypeDecl *Decl,
+                                     DeclContext *Context) {
+    // Early exit if the scope is already fileprivate.
+    if (auto *Nominal = dyn_cast<NominalTypeDecl>(Context))
+      if (isFilePrivate(Nominal))
+        return Parent;
+    if (auto *Func = dyn_cast<AbstractFunctionDecl>(Context))
+      if (isFilePrivate(Func))
+        return Parent;
+
+    // Retrieve the private discriminator.
+    for (; Context->getParent(); Context = Context->getParent())
+      if (auto *SF = dyn_cast<SourceFile>(Context)) {
+        auto PrivateDiscriminator = SF->getPrivateDiscriminator();
+        if (!PrivateDiscriminator.empty())
+          return DBuilder.createNameSpace(Parent, PrivateDiscriminator.str(),
+                                          /*ExportSymbols=*/true);
+    }
+    llvm_unreachable("unknown private discriminator");
+    return Parent;
```

Comment:
`llvm_unreachable` becomes an optimization hint in no-asserts builds, so this `return` is permanently dead.

Reply:
sweet, there is also isInPrivateOrLocalContext() in ASTMangler.